File: 739-daily-temperatures/daily-temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    const n = temperatures.length;
    const result = new Array(n).fill(0);
    const stack = []; // This will store the indices of the temperatures array

    for (let i = 0; i < n; i++) {
        // While the stack is not empty and the current temperature is greater than 
        // the temperature at the index stored at the top of the stack
        while (stack.length > 0 && temperatures[i] > temperatures[stack[stack.length - 1]]) {
            // Pop the index from the stack
            const index = stack.pop();
            // Calculate the number of days until a warmer temperature
            result[index] = i - index; // i is the current day
        }
        // Push the current index onto the stack
        stack.push(i);
    }

    return result;
};

File: 739-daily-temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        daysArr = [0] * len(temperatures)

        currStack = [0]

        for i in range(len(temperatures)):
            while currStack and temperatures[i] > temperatures[currStack[-1]]:
                popped_index = currStack.pop()
                daysArr[popped_index] = i - popped_index
            currStack.append(i)

        return daysArr